File: max_rain_water/max_rain_water.java
public class max_rain_water {
    public static void main(String[] args){
        int[] nums = {2, 8, 1, 4, 6, 2, 1, 8};
        System.out.println(solution(nums));
    }
    private static int solution(int[] height){
        int left = 0; int right = height.length - 1;
        int max = 0;
        for(int i = 0; i < height.length; i++){
            int w = right - left;
            int h = Math.min(height[left], height[right]);
            max = Math.max(max, h*w);
            if(height[left] > height[right]) right--;
            else if(height[right] > height[left]) left++;
            else{left++; right--;}
        } return max;
    }
}

File: max_rain_water/max_rain_water.py
heights = [2, 8, 1, 4, 6, 2, 1, 8]


def solution(heights):
    left = 0
    right = len(heights) - 1
    res = 0

    for i in heights:
        width = right - left
        height = min(heights[left], heights[right])
        res = max(res, width*height)
        if heights[left] > heights[right]:
            right -= 1
        elif heights[right] > heights[left]:
            left += 1
        else:
            right -=1
            left += 1
    return res

print(solution(heights))
    